app/services/notification_service.py

```
@@ -20,67 +24,246 @@
+    def __init__(self, cooldown_minutes: int = 240):
+        self.cooldown = timedelta(minutes=cooldown_minutes)
+        self.last_notification: Dict[str, datetime] = {}
+        self.lock = Lock()
+
+    def should_notify(self, notifier_id: int, event_type: NotificationEventType) -> bool:
+        """Check if notification should be sent based on rate limit."""
+        key = f"{notifier_id}:{event_type.value}"
+
+        with self.lock:
+            last_time = self.last_notification.get(key)
+            if last_time is None or datetime.utcnow() - last_time >= self.cooldown:
```

Ruff (0.14.13)

`datetime.datetime.utcnow()` used

(DTZ003)